File: src/backend/langgraph/adapters/toolAdapter.ts
import type { Tool, ToolProgress } from "src/Tool";
import type { LoopContext } from "src/loopContext";

export interface ToolRun {
  progress: ToolProgress[];
  output: unknown;
  isError: boolean;
}

export async function runTool(
  tool: Tool,
  input: unknown,
  ctx: LoopContext,
): Promise<ToolRun> {
  const progress: ToolProgress[] = [];
  let parsed: unknown = input;
  try {
    parsed = tool.inputSchema.parse(input);
  } catch (err) {
    return {
      progress: [{ type: "progress", msg: `validation failed: ${(err as Error).message}` }],
      output: `validation failed: ${(err as Error).message}`,
      isError: true,
    };
  }
  const gen = tool.call(parsed, ctx);
  let output: unknown;
  while (true) {
    const r = await gen.next();
    if (r.done) {
      output = r.value;
      break;
    }
    progress.push(r.value);
  }
  return { progress, output, isError: false };
}

File: src/docker_agent/engine/adapters/tool_adapter.py
"""Drain a Tool call into a ToolRun.

Parity: ``src/backend/langgraph/adapters/toolAdapter.ts``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from docker_agent.tools.base import Tool, ToolDone, ToolProgress


@dataclass
class ToolRun:
    progress: list[ToolProgress]
    output: Any
    is_error: bool


async def run_tool(tool: Tool[Any, Any], input: Any, ctx: Any) -> ToolRun:
    progress: list[ToolProgress] = []
    gen = tool.call(input, ctx)
    output: Any = None
    async for item in gen:
        if isinstance(item, ToolDone):
            output = item.result
        else:
            progress.append(item)
    return ToolRun(progress=progress, output=output, is_error=False)